src/tree/updater_quantile_hist.cc

```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      temp_qexpand_depth.push_back(ExpandEntry(left_id,
+                                               p_tree->GetDepth(left_id), 0.0, timestamp++));
+      temp_qexpand_depth.push_back(ExpandEntry(right_id,
+                                               p_tree->GetDepth(right_id), 0.0, timestamp++));
+      // - 1 parent + 2 new children
+      num_leaves++;
+    }
+  }
+}
 
-  spliteval_->Reset();
+
+
+void QuantileHistMaker::Builder::ExpandWithDepthWidth(
+  const GHistIndexMatrix &gmat,
+  const GHistIndexBlockMatrix &gmatb,
+  const ColumnMatrix &column_matrix,
+  DMatrix *p_fmat,
+  RegTree *p_tree,
+  const std::vector<GradientPair> &gpair_h) {
+  unsigned timestamp = 0;
+  int num_leaves = 0;
 
   tstart = dmlc::GetTime();
+
   this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
   time_init_data = dmlc::GetTime() - tstart;
 
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1) <<
+    "tree_method=hist does not support multiple roots at this moment";
+
+  // in depth_wise growing, we feed loss_chg with 0.0 since it is not used anyway
+  qexpand_depth_wise_.push_back(ExpandEntry(0, p_tree->GetDepth(0), 0.0, timestamp++));
+  ++num_leaves;
+  tstart = dmlc::GetTime();
+  for (int depth = 0; depth < param_.max_depth + 1; depth++) {
+    int starting_index = std::numeric_limits<int>::max();
+    int sync_count = 0;
+    std::vector<ExpandEntry> temp_qexpand_depth;
+    BuildLocalHistograms(starting_index, sync_count, gmat, gmatb, p_tree, gpair_h);
+    SyncHistograms(starting_index, sync_count, p_tree);
+    time_build_hist += dmlc::GetTime() - tstart;
+    // 2. initNewNode to get it's stats
+    tstart = dmlc::GetTime();
+    BuildNodeStats(gmat, p_fmat, p_tree, gpair_h);
+    time_init_new_node += dmlc::GetTime() - tstart;
+    // 3. evaluateSplit
+    EvaluateSplits(gmat, column_matrix, p_fmat, p_tree, num_leaves, depth, timestamp,
+            temp_qexpand_depth);
+    // 4. if qexpand is empty and temp_qexpand is empty, break
+    if (temp_qexpand_depth.empty()) {
+      qexpand_depth_wise_.clear();
+      nodes_to_derive_.clear();
+      break;
+    } else {
+      qexpand_depth_wise_.clear();
+      nodes_to_derive_.clear();
+      qexpand_depth_wise_ = temp_qexpand_depth;
+      temp_qexpand_depth.clear();
+    }
+  }
+}
+
+void QuantileHistMaker::Builder::ExpandWithLossGuide(
+    const GHistIndexMatrix& gmat,
+    const GHistIndexBlockMatrix& gmatb,
+    const ColumnMatrix& column_matrix,
+    DMatrix* p_fmat,
+    RegTree* p_tree,
+    const std::vector<GradientPair>& gpair_h) {
+
+  unsigned timestamp = 0;
+  int num_leaves = 0;
+
   // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
   CHECK_EQ(p_tree->param.num_roots, 1)
-      << "tree_method=hist does not support multiple roots at this moment";
+    << "tree_method=hist does not support multiple roots at this moment";
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```

Comment:
So `num_roots` is partially supported?  @hcho3 Is there any plan to add full support?  Otherwise we might just remove it.